columnformers/models/layers.py

```diff
@@ -280,3 +283,232 @@ def init_weights(module: nn.Module):
+        # https://pytorch.org/blog/speeding-up-vits/
+        # We were previously using triton blocksparse matmul, but it seems not very
+        # stable. See here:
+        # https://github.com/triton-lang/triton/pull/4156
+        self.weight = nn.Parameter(
+            torch.sparse_bsr_tensor(
+                crow_indices=connectivity.crow_indices(),
+                col_indices=connectivity.col_indices(),
+                values=torch.empty_like(connectivity.values()),
+                size=connectivity.size(),
+            )
+        )
+
+        if bias:
+            self.bias = nn.Parameter(torch.empty(self.out_features))
+        else:
+            self.register_parameter("bias", None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        trunc_normal_(self.weight.values(), std=0.02)
+        if self.bias is not None:
+            nn.init.zeros_(self.bias)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # apply sparse connectivity mask
+        self.weight.values().data.mul_(self.connectivity.values())
+        x = F.linear(x, self.weight, self.bias)
+        return x
+
+    def extra_repr(self) -> str:
+        return (
+            f"{self.in_features}, {self.out_features}, "
+            f"bias={self.bias is not None}, blocksize={self.blocksize}, "
+            f"sparsity={self.sparsity:.2f}"
+        )
+
+
+class BlockSparseLocallyConnected(nn.Module):
+    """
+    A locally connected layer implemented using block sparse linear.
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        kernel_size: int,
+        height: int,
+        depthwise: bool = False,
+        bias: bool = True,
+        blocksize: int = 32,
+        in_shape: Literal["nlc", "nchw"] = "nchw",
+    ):
+        super().__init__()
+        assert isinstance(kernel_size, int), "only square kernels supported"
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.height = height
+        self.kernel_size = kernel_size
+        self.depthwise = depthwise
+        self.blocksize = blocksize
+        self.in_shape = in_shape
+        self.channels_last = not depthwise
+
+        connectivity = _sparse_local_connectivity(
+            in_channels,
+            out_channels,
+            kernel_size,
+            height,
+            depthwise=depthwise,
+            channels_last=self.channels_last,
+        )
+        self.bsl = BlockSparseLinear(
+            connectivity=connectivity, bias=bias, blocksize=blocksize
+        )
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        in_pattern = "n (h w) c" if self.in_shape == "nlc" else "n c h w"
+        out_pattern = "n (h w c)" if self.channels_last else "n (c h w)"
+        output = rearrange(
+            input, f"{in_pattern} -> {out_pattern}", h=self.height, w=self.height
+        )
+        output = self.bsl(output)
+        output = rearrange(
```

Comment:
Does it make sense to keep rearranging the input on each forward pass? Might be slightly more efficient if we assume all inputs are batches of flattened images and then have a separate function to rearrange when necessary (e.g. to visualise activations etc)?

Reply:
Yes good point. If you have `in_shape="nlc"` and `channels_last=True` (or `in_shape="nchw"` and `channels_last=False`) the rearrange will just produce a new view without actually moving the data. Otherwise however, it will move the data a lot like you say.

My thinking was that the rearranging helps the `BlockSparseLocallyConnected` be consistent with `nn.Conv2d` (when `in_shape="nchw"`) and with `nn.Linear` (when `in_shape="nlc"`).

What if we add a third `in_shape` option `nd` that assumes the input is already flattened?

Reply:
Yes I think that's a good idea?